Adding lateral flow test intervention code.

mattea opened the pull request:
Includes base code to test on symptoms and tracing, as well as a number of related parameters.

*Note that the specific parameters (g/b) are currently untuned and not ready for production use.*

This could have a few different alternate implementations, open to tweaks as desired.

Will clean up spacing soon, just wanted to get this out to other eyes.

roberthinch commented on tests/test_interventions.py:
```diff
@@ -2084,6 +2303,186 @@ def test_test_sensitivity(self, test_params ):
+        sd_mult = test_params[ "sd_infectiousness_multiplier" ]
+        sens_exp = test_params[ "lateral_flow_test_sensitivity"]
+        if sd_mult:
+          sens_low = sens_exp - 0.675 / 2.0 * sd_mult
```
Where do these numbers come from? Hard to see from the code in `interventions.c`.

roberthinch commented on tests/test_interventions.py:
```diff
@@ -2084,6 +2303,186 @@ def test_test_sensitivity(self, test_params ):
+
+        # check the sensitivity at the peak.
+        df_test[ "symptomatic" ] = ( df_test["time_symptomatic"] > 0 )
+        true_pos   = df_test[ ( df_test[ "time_since_inf" ] == 3 ) & ( df_test[ "symptomatic" ] == True ) & ( df_test[ "lateral_flow_status" ] == 1 ) & ( df_test[ "test_sensitive_inf" ] == True ) ].groupby( [ "time_since_inf" ] ).size()
```
Why `3`? Would be good if this was not hard-coded but linked back to the code in the model.

roberthinch commented on src/interventions.c:
```diff
@@ -1219,12 +1366,21 @@ void intervention_on_traced(
 
 	parameters *params = model->params;
 
+	int lateral_flow_test = params->lateral_flow_test_on_traced && gsl_ran_bernoulli( rng, params->lateral_flow_test_fraction );
```
If `params->lateral_flow_test_on_traced=FALSE` does the gsl_ran_bernoulli() get called?
Probably best to move to inside the if statement below.

roberthinch commented on src/interventions.c:
```diff
@@ -352,6 +353,15 @@ void update_intervention_policy( model *model, int time )
 		model->manual_trace_notification_quota = params->manual_trace_n_workers * params->manual_trace_notifications_per_worker_day;
 	}
+
+	for( int idx = 0; idx < model->params->n_total; idx++ )
```
I'm not sure whether we should be looping over everybody here, it is very inefficient, especially since this is for an intervention and will only apply to a very small number of people. Can you make it event driven?

roberthinch commented on src/interventions.c:
```diff
@@ -736,6 +746,125 @@ void intervention_test_take( model *model, individual *indiv )
+
+}
+
+/*****************************************************************************************
```
This function does not seem to be used (unless I've missed it's usage). Please remove.

roberthinch commented on src/interventions.c:
```diff
@@ -1140,26 +1287,26 @@ void intervention_on_positive_result( model *model, individual *indiv )
 		time_event = index_token->contact_time + sample_transition_time( model, TEST_RESULT_QUARANTINE );
-		intervention_quarantine_until( model, indiv, NULL, time_event, TRUE, NULL, model->time, 1 );
 	}
+	intervention_quarantine_until( model, indiv, NULL, time_event, TRUE, NULL, model->time, 1 );
```
Shouldn't this remain inside the `if` block?

Also, if the individual does not quarantine, then the household members should not quarantine either.

roberthinch reviewed (changes requested):
Given the size of this change, please can you perform some performance profiling analysis (we really need to add this a reg test as well). There are a few places which looks like it may be adding overhead (even if the feature is turned off), so we need to understand that.

mattea commented:
For some reason I can't respond to some comments, so adding it here:

* I added additional tests for specifically computing the number of tests performed.
* Switched the hardcoded peak to the calculated one.
* I couldn't switch to the new Model system, it broke file writing (e.g. model.write_individual_file())
